src/main/scala/me/rexim/morganey/interpreter/TermOutputHelper.scala

```diff
@@ -1,24 +1,41 @@
 package me.rexim.morganey.interpreter
 
 import me.rexim.morganey.ast.LambdaTerm
-import me.rexim.morganey.church.ChurchNumberConverter._
-import me.rexim.morganey.church.ChurchPairConverter._
+import me.rexim.morganey.meta._
+
+import scala.annotation.tailrec
 
 object TermOutputHelper {
 
-  def isPrintableChar(code: Int): Boolean =
-    32 <= code && code <= 176
-
-  def smartShowTerm(term: LambdaTerm): String = {
-    decodeNumber(term) match {
-      case Some(number) if isPrintableChar(number) => s"char: '${number.toChar}'"
-      case Some(number) => s"number: $number"
-      case None => decodeListOfNumbers(term) match {
-        case Some(numbers) if numbers.forall(isPrintableChar) =>
-          s"""string: \"${numbers.map(_.toChar).mkString}\""""
-        case Some(numbers) => s"numbers: [${numbers.mkString(",")}]"
-        case None => s"term: $term"
+  def smartShowTerm(term: LambdaTerm, noPrefix: Boolean = false): String = {
+    @tailrec
+    def decode(decoders: List[Decoder[_]]): String = decoders match {
+      case hd :: tl => hd(term) match {
+        case Some(result) => if (noPrefix) result else s"${hd.prefix}: $result"
+        case None         => decode(tl)
       }
+      case Nil =>
+        // `idDecoder` won't ever return `None`
+        val result = idDecoder(term).get
+        if (noPrefix) result else s"${idDecoder.prefix}: $result"
     }
+    decode(decoders)
+  }
+
+  private implicit val idDecoder = Decoder[LambdaTerm]("term", _.toString)
```

Comment:
I don't know why, but this instance is `implicit`. The `implicit` keyword could be removed. This might be a autocomplete-related issue with my IDE.